### PR TITLE
Update Advanced Preprocessing for Algo Trading.py

### PR DESCRIPTION
**Pull request by raghavdixit99**
---
# EntelligenceAI PR Summary 
 **Purpose**:
- The updated script refines our algorithmic trading strategy with advanced signal generation, execution strategies, and comprehensive performance metrics for enhanced decision-making and risk management.

**Changes**:
- **Enhancement**: Improved `Advanced Preprocessing for Algo Trading.py` with stop-loss features, comprehensive performance metrics (cumulative return, Sharpe ratio, Sortino ratio, win rate), and dynamic market condition-based signals for flexible strategy execution.

**Impact**:
- Elevates user experience by enhancing navigation and adaptability in varying market conditions, improving risk management and decision-making, and ensuring robustness and reliability of the trading strategy.

### CODE REVIEW

File: Advanced Preprocessing for Algo Trading.py [hunk 1]
```diff
@@ -1,61 +1,99 @@
...
+                data.loc[i, "Strategy_Returns"] = net_return
+                stop_loss, take_profit, trailing_stop_level = None, None, None
+
+        if current_signal == "Buy":
+            data.loc[i, "Position"] = 1
+        elif current_signal == "Sell":
+            data.loc[i, "Position"] = 0
+            stop_loss, take_profit, trailing_stop_level = None, None, None
+        else:
+            data.loc[i, "Position"] = data.loc[i - 1, "Position"]
```
Comment (raghavdixit99) [inline Advanced Preprocessing for Algo Trading.py:63]: Strategy returns are only calculated on stop loss hits, missing returns calculation for normal position exits on `Sell` signals which leads to incomplete performance metrics.

File: Advanced Preprocessing for Algo Trading.py [hunk 2]
```diff
@@ -1,61 +1,99 @@
...
+            if stop_loss is None:
+                stop_loss = data.loc[i - 1, "Close"] - atr_multiplier * atr
+                take_profit = data.loc[i - 1, "Close"] + atr_multiplier * atr
+                trailing_stop_level = stop_loss if trailing_stop else None
+
+            if trailing_stop and data.loc[i, "Close"] > take_profit * 0.5:
+                trailing_stop_level = max(trailing_stop_level, data.loc[i, "Close"] - atr)
+
+            if data.loc[i, "Close"] <= stop_loss or (trailing_stop and data.loc[i, "Close"] <= trailing_stop_level):
+                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
```
Comment (raghavdixit99) [inline Advanced Preprocessing for Algo Trading.py:51]: Missing handling of take-profit levels in strategy execution. Take-profit targets are calculated but never used to exit positions.

File: Advanced Preprocessing for Algo Trading.py [hunk 3]
```diff
@@ -1,61 +1,99 @@
...
+                net_return = gross_return - slippage - transaction_cost
+                data.loc[i, "Strategy_Returns"] = net_return
```
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:53]: Strategy returns are only calculated on stop loss hits, missing returns from profitable trades and normal exits. Need to calculate returns for all position changes.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Position"] != data.loc[i - 1, "Position"]:
                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
                net_return = gross_return - slippage - transaction_cost
                data.loc[i, "Strategy_Returns"] = net_return
```
</details>

File: Advanced Preprocessing for Algo Trading.py [hunk 4]
```diff
@@ -1,61 +1,99 @@
...
+        ranging_mask & (data["Close"] > data["BB_Upper"]) & (data["Volatility_Breakout"] == 1), "Signal"] = "Sell"
+
+    data.loc[choppy_mask & (data["Volatility_Breakout"] == 1), "Signal"] = "Buy"
+
+    return data
+
+def execute_strategy_with_stop_loss(df, slippage=0.0002, transaction_cost=0.001, atr_multiplier=2.5, trailing_stop=True):
     data = df.copy()
+    data["Position"] = 0
+    data["Strategy_Returns"] = 0.0
```
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:33]: Missing initialization of `Position` for first row (index 0), which could cause reference errors when accessing `data.loc[i-1, "Position"]`


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
    data = df.copy()
    data["Position"] = 0
    data.loc[0, "Position"] = 0
    data["Strategy_Returns"] = 0.0
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 3 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:53]: Strategy returns are only calculated on stop loss hits, missing returns calculation for normal position exits via `Sell` signals which can lead to incorrect performance metrics.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Close"] <= stop_loss or (trailing_stop and data.loc[i, "Close"] <= trailing_stop_level) or data.loc[i, "Signal"] == "Sell":
                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
                net_return = gross_return - slippage - transaction_cost
                data.loc[i, "Strategy_Returns"] = net_return
```
</details>

File: Advanced Preprocessing for Algo Trading.py [hunk 5]
```diff
@@ -1,61 +1,99 @@
...
+            if data.loc[i, "Close"] <= stop_loss or (trailing_stop and data.loc[i, "Close"] <= trailing_stop_level):
+                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
+                net_return = gross_return - slippage - transaction_cost
+                data.loc[i, "Strategy_Returns"] = net_return
+                stop_loss, take_profit, trailing_stop_level = None, None, None
+
+        if current_signal == "Buy":
+            data.loc[i, "Position"] = 1
+        elif current_signal == "Sell":
+            data.loc[i, "Position"] = 0
```
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:59]: Missing initialization of `Strategy_Returns` for buy entries, which should include transaction costs and slippage, leading to underestimated trading costs.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
if current_signal == "Buy":
            data.loc[i, "Position"] = 1
            data.loc[i, "Strategy_Returns"] = data.loc[i, "Returns"] - (transaction_cost + slippage)
        elif current_signal == "Sell":
            data.loc[i, "Position"] = 0
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 3 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:53]: Strategy returns are only calculated on stop loss hits, missing returns from profitable trades and normal exits. Need to calculate returns for all position changes.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Position"] != data.loc[i - 1, "Position"]:
                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
                net_return = gross_return - slippage - transaction_cost
                data.loc[i, "Strategy_Returns"] = net_return
```
</details>

File: Advanced Preprocessing for Algo Trading.py [hunk 6]
```diff
@@ -1,61 +1,99 @@
...
-    # Calculate Volatility Breakout
-    data["Volatility_Breakout"] = np.where(data["High"] - data["Low"] > 1.5 * atr, 1, 0)
+        if previous_signal == "Buy":
+            if stop_loss is None:
+                stop_loss = data.loc[i - 1, "Close"] - atr_multiplier * atr
+                take_profit = data.loc[i - 1, "Close"] + atr_multiplier * atr
+                trailing_stop_level = stop_loss if trailing_stop else None
+
+            if trailing_stop and data.loc[i, "Close"] > take_profit * 0.5:
+                trailing_stop_level = max(trailing_stop_level, data.loc[i, "Close"] - atr)
```
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:48]: Missing position exit logic when take profit level is hit. Strategy continues holding position even after reaching take profit target.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Close"] >= take_profit:
                position = 0
                trailing_stop_level = 0
            elif trailing_stop and data.loc[i, "Close"] > take_profit * 0.5:
                trailing_stop_level = max(trailing_stop_level, data.loc[i, "Close"] - atr)
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 3 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:53]: Strategy returns are only calculated on stop loss hits, missing returns from profitable trades and normal exits. Need to calculate returns for all position changes.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Position"] != data.loc[i - 1, "Position"]:
                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
                net_return = gross_return - slippage - transaction_cost
                data.loc[i, "Strategy_Returns"] = net_return
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 4 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:33]: Missing initialization of `Position` for first row (index 0), which could cause reference errors when checking previous positions.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
    data = df.copy()
    data["Position"] = 0
    data.loc[0, "Position"] = 0
    data["Strategy_Returns"] = 0.0
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 3 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:53]: Strategy returns are only calculated on stop loss hits, missing returns calculation for normal position exits on `Sell` signals which will underreport performance metrics.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
            if data.loc[i, "Close"] <= stop_loss or (trailing_stop and data.loc[i, "Close"] <= trailing_stop_level) or data.loc[i, "Signal"] == "Sell":
                gross_return = data.loc[i, "Close"] / data.loc[i - 1, "Close"] - 1
                net_return = gross_return - slippage - transaction_cost
                data.loc[i, "Strategy_Returns"] = net_return
```
</details>

File: Advanced Preprocessing for Algo Trading.py (hunk 4 above)
Comment (aiswaryasankarwork) [inline Advanced Preprocessing for Algo Trading.py:33]: Missing initialization of `Position` for first row (index 0) which could cause incorrect position tracking when accessing `data.loc[i-1, "Position"]`.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation. 


```suggestion
    data = df.copy()
    data["Position"] = 0
    data.loc[0, "Position"] = 0
    data["Strategy_Returns"] = 0.0
```
</details>
